staging/src/k8s.io/apiserver/pkg/cel/library/network_addrs.go

```diff
@@ -0,0 +1,141 @@
+limitations under the License.
+*/
+
+package library
+
+import (
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// NetworkAddr provides a CEL function library extension of network address parsing functions.
+//
+// networkAddr
+//
+// Converts a string to a network address or results in an error if the string is not a valid network address.
+// The network address must be an IPv4 or IPv6 address with a CIDR mask.
+//
+//	networkAddr(<string>) <NetworkAddr>
+//
+// Examples:
+//
+//	networkAddr('192.168.0.0/16') // returns an IPv4 address with a CIDR mask
+//	networkAddr('::1/128') // returns an IPv6 address with a CIDR mask
+//	networkAddr('192.168.0.0/33') // error
+//	networkAddr('::1/129') // error
+//
+// isNetworkAddr
+//
+// Returns true if a string is a valid network address. The network address must be an IPv4 or IPv6 address with a CIDR mask.
+//
+//	isNetworkAddr(<string>) <bool>
+//
+// Examples:
+//
+//	isNetworkAddr('192.168.0.0/16') // returns true
+//	isNetworkAddr('::1/128') // returns true
+//	isNetworkAddr('192.168.0.0/33') // returns false
+//	isNetworkAddr('::1/129') // returns false
+//
+// containsIPAddr
+//
+// - containsIPAddr: Returns true if a network address contains an IP address. Should be passed an IP address parsed with ipAddr.
+//
+// Examples:
+//
+// networkAddr('192.168.0.0/24').containsIPAddr(ipAddr('192.168.0.1')) // returns true
+// networkAddr('192.168.0.0/24').containsIPAddr(ipAddr('192.168.1.1')) // returns false
+func NetworkAddrs() cel.EnvOption {
+	return cel.Lib(networkAddrsLib)
+}
+
+var networkAddrsLib = &networkAddrs{}
+
+type networkAddrs struct{}
+
+func (*networkAddrs) LibraryName() string {
+	return "kubernetes.networkaddrs"
+}
+
+var networkAddrLibraryDecls = map[string][]cel.FunctionOpt{
+	"networkAddr": {
+		cel.Overload("string_to_network_addr", []*cel.Type{cel.StringType}, apiservercel.NetworkAddrType,
+			cel.UnaryBinding(stringToNetworkAddr)),
+	},
+	"containsIPAddr": {
+		cel.MemberOverload("network_addr_contains_ip_addr", []*cel.Type{apiservercel.NetworkAddrType, apiservercel.IPAddrType}, cel.BoolType,
+			cel.BinaryBinding(containsIPAddr)),
+	},
+	"isNetworkAddr": {
+		cel.Overload("is_network_addr", []*cel.Type{cel.StringType}, cel.BoolType,
+			cel.UnaryBinding(isNetworkAddr)),
+	},
+}
+
+func (*networkAddrs) CompileOptions() []cel.EnvOption {
+	options := []cel.EnvOption{}
+	for name, overloads := range networkAddrLibraryDecls {
+		options = append(options, cel.Function(name, overloads...))
+	}
+	return options
+}
+
+func (*networkAddrs) ProgramOptions() []cel.ProgramOption {
+	return []cel.ProgramOption{}
+}
+
+func stringToNetworkAddr(arg ref.Val) ref.Val {
+	s, ok := arg.Value().(string)
+	if !ok {
+		return types.MaybeNoSuchOverloadErr(arg)
+	}
+
+	net, err := netip.ParsePrefix(s)
```

Comment:
[My version](https://github.com/danwinship/kubernetes/commit/9bc20ff8cc4ff75d41b4e256889ba1b8c8134828#diff-f4b3c1b088debfb8229ad7432d5f86310c531d3c9c4a2e8f3cdba9bf665f01c0R50-R73). In particular:

1. Definitely don't allow 4-in-6 (e.g. "`::ffff:1.2.3.0/24`"). (`net.ParseCIDR` and `netip.ParsePrefix` have incompatible ideas on how to parse 4-in-6 CIDRs anyway.)
2. Require that `net == net.Masked()`, to limit it to the subnet/mask case rather than the IP-with-prefix-length case.

(The code there also has another check to rule out the current golang stdlib wackiness Antonio mentioned, which will be fixed in golang 1.22, plus another error-message-wrapper.)